File: viewclick/src/main/java/com/max/viewclick/ViewClickManager.java
package com.max.viewclick;

import android.os.Handler;
import android.os.Message;
import android.view.View;

import java.util.ArrayList;
import java.util.HashMap;

/**
 * Created by max on 2017/11/21.
 *
 * @email maxzhou7@gmail.com
 */

public class ViewClickManager implements Handler.Callback {

    public interface OnClickListener {
        void onSingleClick();

        void onDoubleClick();
    }

    private static final int INTERVAL = 300;//双击间隔时间 毫秒
    private static final int WHAT_CLICK = 1000;

    private HashMap<Integer, ArrayList<OnClickListener>> mViews;//save the clickable view，and last click time fo views
    private Handler mHandle;

    public ViewClickManager() {
        //init somethings
        mViews = new HashMap<>();
        mHandle = new Handler(this);
    }

    @Override
    public boolean handleMessage(Message msg) {
        if (msg.what == WHAT_CLICK) {
            ArrayList<OnClickListener> tt = mViews.get(msg.arg1);
            if (tt == null) return true;
            if (tt != null && tt.size() >= 2) {
                //double click
                tt.get(0).onDoubleClick();
            } else {
                tt.get(0).onSingleClick();
            }
            mViews.remove(msg.arg1);
        }
        return false;
    }

    public void setClickLister(final View view, final OnClickListener lister) {
        if (view == null || lister == null) {
            throw new NullPointerException("view or lister is null");
        }
        final int id = view.getId();
        if (mViews.containsKey(id)) mViews.remove(id);
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ArrayList<OnClickListener> tempList;
                if (mViews.containsKey(id)) {
                    tempList = mViews.get(id);
                } else {
                    // this is first click
                    tempList = new ArrayList<>();

                    Message message = mHandle.obtainMessage(WHAT_CLICK);
                    message.arg1 = id;
                    mHandle.sendMessageDelayed(message, INTERVAL);
                }
                tempList.add(lister);
                mViews.put(id, tempList);
            }
        });
    }
}

File: utilslib/src/main/java/com/max/utils/ViewClickManager.kt
package com.max.utils

import android.os.Handler
import android.os.Message
import android.view.View
import java.util.*

/**
 * Created by max on 2017/11/22.
 */
class ViewClickManager : Handler.Callback {

    private val INTERVAL = 300//双击间隔时间 毫秒
    private val WHAT_CLICK = 1000

    //save the clickable view，and last click time fo views
    private val mViews = HashMap<Int, ArrayList<OnClickListener>>()
    private val mHandle = Handler(this)

    override fun handleMessage(msg: Message?): Boolean {
        when (msg?.what) {
            WHAT_CLICK -> {
                var temp = mViews[msg?.arg1] ?: return true
                if (temp != null && temp.size >= 2) {
                    //double click
                    temp[0].onDoubleClick()
                } else {
                    temp[0].onSingleClick()
                }
                mViews.remove(msg.arg1)
            }
        }
        return false
    }

    fun setClickLister(view: View?, lister: OnClickListener?) {
        if (view == null || lister == null) {
            throw NullPointerException("view or lister is null")
        }
        val id = view.id
        if (mViews.containsKey(id)) mViews.remove(id)
        view.setOnClickListener {
            val tempList: ArrayList<OnClickListener>
            if (mViews.containsKey(id)) {
                tempList = mViews[id] ?: ArrayList()
            } else {
                // this is first click
                tempList = ArrayList()

                val message = mHandle.obtainMessage(WHAT_CLICK)
                message.arg1 = id
                mHandle.sendMessageDelayed(message, INTERVAL.toLong())
            }
            tempList.add(lister)
            mViews.put(id, tempList)
        }
    }

    interface OnClickListener {
        fun onSingleClick()

        fun onDoubleClick()
    }
}